### ISSUE TITLE
how to apply this accuracy to the GO/NG of cooling fan AOI inspection ?

### ISSUE THREAD

**Issue by Austin-Chang-zz** [Austin-Chang-zz/Fan_AOI_Austin#5]
Great question! When using a **Convolutional Neural Network (CNN)** for **Automatic Optical Inspection (AOI)** in a production line, such as inspecting the labels on cooling fans, you can leverage the model's accuracy to determine whether a product passes the inspection (GO) or fails it (NO GO). Let's break down the process of applying CNN accuracy to the GO/NG decision-making for your specific use case.

#### 1. **Define the Problem and Dataset**
Your goal is to inspect the cooling fan labels to determine if they are aligned correctly:
- **GO**: The label is correctly aligned (within an acceptable angle).
- **NO GO**: The label is tilted by more than 15 degrees or has other defects.

You have already labeled your dataset, where images with properly aligned labels are marked as "GO" (e.g., `100`), and images with tilted or defective labels are marked as "NG" (e.g., `60`). 

#### 2. **Train a CNN Model for Classification**
You can train a CNN model to classify images into two categories:
- **Class 0**: NG (Label is tilted or defective)
- **Class 1**: GO (Label is correctly aligned)

The model's output will be a **probability score** for each class.

Here’s a simplified example of how you can train a CNN using TensorFlow/Keras:

```python
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense
from tensorflow.keras.preprocessing.image import ImageDataGenerator

# Define the CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(80, 80, 3)),
    MaxPooling2D(2, 2),
    Conv2D(64, (3, 3), activation='relu'),
    MaxPooling2D(2, 2),
    Flatten(),
    Dense(128, activation='relu'),
    Dense(2, activation='softmax')  # Output layer for binary classification (GO/NG)
])

# Compile the model
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

# Assume you have train and validation data in ImageDataGenerator format
train_datagen = ImageDataGenerator(rescale=1./255)
train_generator = train_datagen.flow_from_directory(
    'C:/Photos/train',
    target_size=(80, 80),
    batch_size=32,
    class_mode='binary'
)

# Train the model
model.fit(train_generator, epochs=10)
```

#### 3. **Model Evaluation: Get the Accuracy**
- After training, the model will produce an **accuracy metric** based on how well it classifies the test data.
- A higher accuracy means that the model is better at distinguishing between GO and NG samples.

#### 4. **Use CNN Model for GO/NG Decision**
Now, let's apply the trained CNN model to classify new cooling fan images in real-time on the production line:

```python
import cv2
import numpy as np

# Load the pre-trained model
model = tf.keras.models.load_model("C:/Photos/fan_model.h5")

# Function to classify a single image
def classify_image(image_path):
    img = cv2.imread(image_path)
    img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
    img = cv2.resize(img, (80, 80))
    img = img.reshape(1, 80, 80, 3) / 255.0  # Normalize
    prediction = model.predict(img)
    # Get the predicted class (0 for NG, 1 for GO)
    class_idx = np.argmax(prediction, axis=1)[0]
    confidence = prediction[0][class_idx] * 100  # Confidence score in percentage
    return class_idx, confidence

# Example usage
image_path = "C:/Photos/sample_fan.jpg"
result, confidence = classify_image(image_path)
if result == 1:
    print(f"GO (Correctly Aligned) with {confidence:.2f}% confidence")
else:
    print(f"NG (Defective) with {confidence:.2f}% confidence")
```

#### Explanation:
- The **`model.predict()`** function returns probabilities for each class.
- **`np.argmax()`** selects the class with the highest probability.
- If the predicted class is **1**, it means the label is aligned (GO); otherwise, it's tilted (NG).

#### 5. **Setting a Threshold for NG Criteria**
Sometimes, you may want to be more stringent about what counts as a GO:
- You can set a **confidence threshold**. For example, if the model predicts `GO` with less than 90% confidence, you can still classify it as `NG`.

```python
if result == 1 and confidence >= 90:
    print("GO")
else:
    print("NG")
```

### 6. **Deploying the Model on the Production Line**
To automate the GO/NG decision in real-time:
- Use a camera to capture images of each cooling fan.
- Apply the trained CNN model to each image.
- Trigger an **audio alarm** or **stop the production line** if an NG case is detected.

### Summary
1. **Train a CNN model** to classify GO vs. NG labels.
2. **Use the model's predictions** (accuracy and confidence score) to determine GO/NG in real-time.
3. Set a **confidence threshold** to reduce false positives or negatives.
4. **Deploy the solution** on your production line for automated inspections.

Would you like more details on any specific part, or do you need additional help with implementation?
